project-example/src/main.rs

```diff
@@ -44,40 +48,188 @@ fn main() {
+                // component for the purpose of this example.
+                //
+                // NOTE: This assumes that the entity is already present in `world`. This
+                // is a safe assumption to make because SpatialOS will always notify us that
+                // of a new entity before sending any component data for that entity.
                 WorkerOp::AddComponent(add_component) => match add_component.component_id {
-                    example::Example::ID => {
-                        let component_data = add_component.get::<example::Example>().unwrap();
-                        println!("Received Example data: {:?}", component_data);
+                    example::Rotate::ID => {
+                        let rotate = add_component.get::<example::Rotate>().unwrap();
+                        let entity_state = world
+                            .get_mut(&add_component.entity_id)
+                            .expect("Entity wasn't present in local world");
+                        entity_state.rotate = Some(rotate.clone());
                     }
                     id => println!("Received unknown component: {}", id),
                 },
+
+                // Track authority changes to the `Rotate` component. We only want to update
+                // entities where we're authoritative over the `Rotate` component, so we
+                // need to know which entities have authoritative over.
+                WorkerOp::AuthorityChange(authority_change) => {
+                    if authority_change.component_id == example::Rotate::ID {
+                        println!(
+                            "Authority change for {}: {:?}",
+                            c.get_worker_id(),
+                            authority_change
+                        );
+                        let state = world.get_mut(&authority_change.entity_id).unwrap();
+                        state.has_authority = authority_change.authority.has_authority();
+                    }
+                }
+
+                // Update the locally tracked state of a component. We override any changes
+                // that we have made locally with the data sent by the server in order to
+                // ensure that we're always working off the latest canonical state of the
+                // component.
                 WorkerOp::ComponentUpdate(update) => match update.component_id {
-                    example::Example::ID => {
-                        let component_update = update.get::<example::Example>();
-                        println!("Received Example update: {:?}", component_update)
+                    example::Rotate::ID => {
+                        let component_update = update.get::<example::Rotate>().unwrap();
+                        let state = world.get_mut(&update.entity_id).unwrap();
+                        let rotate = state.rotate.as_mut().unwrap();
+                        rotate.merge(component_update.clone());
```

Comment:
Should the `merge` method take the update by reference rather than value?

Reply:
I'm fine with it taking the update by value. If merge were to take the update object by reference, then it would need to copy the individual fields in order to apply them to the component. Since it ultimately needs the component data by value, it makes more sense to take the update by value, that way the calling code can determine if the value needs to be cloned. This seems to be the more common convention in Rust (i.e. let the calling code decide whether or not the clone the data).